File: events/server/Sr.Grpc/EventServer/WeatherImpl.cs
using Grpc.Core;
using Sr.Grpc.gen;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading.Tasks;

namespace server.Sr.Grpc.EventServer;

public class WeatherImpl : WeatherSubscriber.WeatherSubscriberBase
{
    private readonly object _lock = new();
    private readonly Random _generator = new();
    
    private readonly List<WeatherEvent> _events = [];

    private readonly List<string> _locations =
    [
        "Cracow", "Warsaw", "Rzeszow", "Gdansk", "Katowice"
    ];
    
    private WeatherEvent _GenerateWeatherEvent()
    {
        List<WeatherType> values = [WeatherType.Rain, WeatherType.Storm, WeatherType.Sun,
            WeatherType.Clouds, WeatherType.Fog];
        var type = values[_generator.Next(0, values.Count)];
        var location = _locations[_generator.Next(0, _locations.Count)];
        var temperature = -10.0f + (float)_generator.NextDouble() * 44.0f;
        return new WeatherEvent
        {
            Location = location,
            Temperature = temperature,
            Type = type
        };
    }
    
    public void GenerateEvents()
    {
        lock (_lock)
        {
            _events.Clear();
            var eventsNumber = _generator.Next(1, 5);
            for (var i = 0; i < eventsNumber; i++)
            {
                _events.Add(_GenerateWeatherEvent());
            }    
        }
    }

    public override async Task Subscribe(WeatherSubscription request, IServerStreamWriter<WeatherEvent> responseStream,
        ServerCallContext context)
    {
        Console.WriteLine("Weather: A subscription started");
        WeatherEvent? lastEvent = null;
        try
        {
            while (true)
            {
                if (context.CancellationToken.IsCancellationRequested)
                {
                    Console.WriteLine("Weather: a client cancelled");
                    return;
                }

                WeatherEvent? toSend = null;
                lock (_lock)
                {
                    var found = _events.Find(weatherEvent => request.Location == weatherEvent.Location);
                    if (found is not null)
                    {
                        toSend = found;
                    }
                }

                if (toSend is not null && !toSend.Equals(lastEvent))
                {
                    await responseStream.WriteAsync(toSend);
                }
                
                lastEvent = toSend;
                await Task.Delay(TimeSpan.FromSeconds(_generator.Next(1, 3)));

            }

        }
        catch (RpcException ex) when (ex.StatusCode == StatusCode.Cancelled)
        {
            Console.WriteLine("Weather: a client cancelled");
        }
        catch (RpcException ex) when (ex.StatusCode != StatusCode.Cancelled)
        {
            Console.WriteLine($"Weather: an error occured: {ex.StatusCode}");
        }

        Console.WriteLine("Weather: subscription end");
    }
}

File: events/server/weather_impl.py
import asyncio
import random

import generated.weather_pb2_grpc as weather_grpc
import generated.weather_pb2 as weather


class WeatherImpl(weather_grpc.WeatherSubscriberServicer):
    def __init__(self):
        self._lock = asyncio.Lock()
        self._events = []
        self._locations = ["Cracow", "Warsaw", "Rzeszow", "Gdansk", "Katowice"]

    def _generate_weather_event(self):
        weather_types = [
            weather.WeatherType.RAIN,
            weather.WeatherType.STORM,
            weather.WeatherType.SUN,
            weather.WeatherType.CLOUDS,
            weather.WeatherType.FOG
        ]
        event = weather.WeatherEvent(
            location=random.choice(self._locations),
            temperature=random.uniform(-10.0, 34.0),
            type=random.choice(weather_types)
        )

        return event

    def generate_events(self):
        events = [self._generate_weather_event() for _ in range(random.randint(1, 4))]
        asyncio.create_task(self._update_events(events))

    async def _update_events(self, events):
        async with self._lock:
            self._events = events

    async def Subscribe(self, request, context):
        print("Weather: a subscription started")
        last_event = None

        try:
            while not context.done():
                to_send = None
                async with self._lock:
                    for e in self._events:
                        if e.location == request.location:
                            to_send = e
                            break

                if to_send and (last_event != to_send):
                    await context.write(to_send)
                    last_event = to_send

                await asyncio.sleep(random.randint(1, 2))
        except asyncio.CancelledError:
            print("Weather: a client cancelled")
        except Exception as ex:
            print(f"Weather: an error occurred: {ex}")
        print("Weather: subscription end")
